compiler/src/lnntoamm/Fn.ts

```diff
@@ -112,90 +102,90 @@ export default class Fn {
+        throw new Error(`event handlers should not return values`);
       }
       amm.exit();
     }
   }
 
-  // TODO: this will have to change in order to call fns multiple times - maybe deep cloning?
-  inline(amm: Output, args: Ref[], assign: string, kind: 'const' | 'let' | '') {
-    let argNames = Object.keys(this.args);
-    if (argNames.length !== args.length) {
-      // this should be caught by Call, it's just a sanity check
-      throw new Error(`number of arguments off`);
-    }
-    for (let ii = 0; ii < argNames.length; ii++) {
-      this.args[argNames[ii]].val = args[ii];
-    }
-    const last = this.body[this.body.length - 1];
-    if (last instanceof Exit && last.exitVal !== null) {
-      if (kind === 'const' && last.exitVal.dec.mutable) {
-        kind = 'let';
-      } else if (kind === 'let' && !last.exitVal.dec.mutable) {
-        last.exitVal.dec.mutable = true;
-      } else if (kind === '') {
-        TODO('figure out how to do return value rewrites pls');
-      }
-      last.exitVal.dec.ammName = assign;
+  // FIXME: it'll take a bit more work to do better inlining, but it *should* be possible
+  // to have `inline` load all of the amm code into a new `Stmt[]` which is then iterated
+  // at the handler level to do optimizations and such, similar to the `Microstatement[]`
+  // that was loaded but using the same JS objects(?) and the optimizations should only
+  // be further inlining...
+  // FIXME: another option is to convert to SSA form (talked a bit about in Amm.ts) and then
+  // perform optimizations from there. This *might* require the `Stmt[]` array from above
+  // *or* we can do it in Amm.ts using only strings (although that might be harder)
+  // FIXME: a 3rd option is to make amm itself only SSA and perform the the "register
+  // selection" in the ammtox stage. This might be the best solution, since it's the most
+  // flexible regardless of the backend, and amm is where that diverges.
+  inline(amm: Output, args: Ref[], kind: AssignKind, name: string, ty: Builtin) {
+    if (args.length !== this.params.length) {
+      throw new Error(`function call argument mismatch`);
     }
+    this.params.forEach((param, ii) => param.assign(args[ii]));
     for (let ii = 0; ii < this.body.length; ii++) {
       const stmt = this.body[ii];
-      if (stmt instanceof Dec || stmt instanceof Assign || stmt instanceof Emit) {
-        stmt.inline(amm);
-      } else if (stmt instanceof Exit) {
-        amm.assign(
-          kind,
-          assign,
-          stmt.exitVal.ty.breakdown(),
-          stmt.exitVal.ammName,
-        );
-      } else {
-        throw new Error(`did not expect to inline stmt: ${stmt}`);
+      if (stmt instanceof Exit) {
+        if (ii !== this.body.length - 1) {
+          throw new Error(`got a return at a bad time (should've been caught already?)`);
+        }
+        let refCall = 'reff';
+        if (ty.eq(opcodes().get('string'))) {
+          refCall = 'refv';
+        }
```

Comment:
Maybe a `TODO` that this will eventually impact all other types `Array`, `HashMap`, `Result`, etc, and that this conditional should be inverted to something like:

```js
let refCall = 'refv';
if (['int8', 'int16', 'int32', 'int64', 'float32', 'float64', 'bool'].includes(ty.name)) {
  refCall = 'reff';
}
```

Reply:
i'll just go ahead and fix it now :)